### ISSUE TITLE
[bug] 'PermissionError: [WinError 32]' during parallel execution of 'conan export-pkg'

### ISSUE THREAD

**Issue by DmitrySokolov** [conan-io/conan#6864]
Conanfiles:
```python
# testA/conanfile.py

from conans import ConanFile
from os.path import join
from time import sleep

class TestA(ConanFile):
    name = "TestA"
    version = "1.0.0"
    description = "TestA"
    exports = "*.a"

    def package(self):
        super().package()
        self.copy("*.a", src=".", dst="lib")
        with open(join(self.package_folder, "config.txt"), "wt") as fh:
            sleep(10)
            fh.write("testA\n")

    def package_info(self):
        super().package_info()
        self.cpp_info.libs = ["lib1", "lib2", "lib3"]


# testB/conanfile.py

from conans import ConanFile
from time import sleep

class TestB(ConanFile):
    name = "TestB"
    version = "1.0.0"
    description = "TestB"
    exports = "*.a"

    requires = "TestA/1.0.0@test/develop"

    def configure(self):
        super().configure()
        sleep(5)

    def package(self):
        super().package()
        self.copy("*.a", src=".", dst="lib")

    def package_info(self):
        super().package_info()
        self.cpp_info.libs = ["lib4"]
```

create.cmd
```cmd
set CONAN_LOGGING_LEVEL=10

conan remove "test*" -f
conan export testA test/develop

echo/
pushd testA
start "A" cmd /k conan export-pkg . test/develop
popd

echo/
pushd testB
start "B" cmd /k conan export-pkg . test/develop
popd
```

Output of testA
```cmd
DEBUG :conan_api.py   [163]: INIT: Using config 'C:\Users\dmitry\.conan\conan.conf' [2020-04-15 23:43:55,615]
DEBUG :tracer.py      [156]: CONAN_API: export_pkg(conanfile_path=.,name=None,version=None,source_folder=None,build_folder=None,package_folder=None,install_folder=None,profile_names=None,env=None,settings=None,options=None,force=False,user=test,channel=develop,lockfile=None,ignore_dirty=False) [2020-04-15 23:43:55,617]
DEBUG :profile_loader.py[112]: PROFILE LOAD: C:\Users\dmitry\.conan\profiles\default [2020-04-15 23:43:55,618]
INFO  :settings_preprocessor.py[58]: Setting 'compiler.runtime' not declared, automatically adjusted to 'MD' [2020-04-15 23:43:55,639]
[HOOK - attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'url'. It is recommended to add it as attribute
[HOOK - attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'license'. It is recommended to add it as attribute
DEBUG :export.py      [105]: EXPORT: C:\Dev\Projects\tests\conan_export_pkg\testA\conanfile.py [2020-04-15 23:43:55,902]
Exporting package recipe
TestA/1.0.0@test/develop exports: Copied 3 '.a' files: liblib1.a, liblib2.a, liblib3.a
TestA/1.0.0@test/develop: The stored package has not changed
TestA/1.0.0@test/develop: Exported revision: 4dbefb4080a01fd913f3844b071ac62e
DEBUG :graph_builder.py[391]: GRAPH: new_node: TestA/1.0.0@test/develop [2020-04-15 23:43:55,922]
DEBUG :graph_builder.py[66]: GRAPH: Time to load deps 0.004001140594482422 [2020-04-15 23:43:55,922]
TestA/1.0.0@test/develop: Forced build from source
Packaging to 5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
TestA/1.0.0@test/develop: Generating the package
TestA/1.0.0@test/develop: Package folder C:\Users\dmitry\.conan\data\TestA\1.0.0\test\develop\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
TestA/1.0.0@test/develop: Calling package()
TestA/1.0.0@test/develop: WARN: This conanfile has no package step
DEBUG :packager.py    [120]: PACKAGE: Creating config files to C:\Users\dmitry\.conan\data\TestA\1.0.0\test\develop\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 [2020-04-15 23:44:05,928]
TestA/1.0.0@test/develop package(): Packaged 1 '.txt' file: config.txt
TestA/1.0.0@test/develop package(): Packaged 3 '.a' files: liblib1.a, liblib2.a, liblib3.a
TestA/1.0.0@test/develop: Package '5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9' created
TestA/1.0.0@test/develop: Created package revision 9c1ad6891f2902de5e6eae0d6d39da9e
```

Output of testB:
```cmd
DEBUG :conan_api.py   [163]: INIT: Using config 'C:\Users\dmitry\.conan\conan.conf' [2020-04-15 23:43:55,615]
DEBUG :tracer.py      [156]: CONAN_API: export_pkg(conanfile_path=.,name=None,version=None,source_folder=None,build_folder=None,package_folder=None,install_folder=None,profile_names=None,env=None,settings=None,options=None,force=False,user=test,channel=develop,lockfile=None,ignore_dirty=False) [2020-04-15 23:43:55,617]
DEBUG :profile_loader.py[112]: PROFILE LOAD: C:\Users\dmitry\.conan\profiles\default [2020-04-15 23:43:55,618]
INFO  :settings_preprocessor.py[58]: Setting 'compiler.runtime' not declared, automatically adjusted to 'MD' [2020-04-15 23:43:55,638]
[HOOK - attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'url'. It is recommended to add it as attribute
[HOOK - attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'license'. It is recommended to add it as attribute
DEBUG :export.py      [105]: EXPORT: C:\Dev\Projects\tests\conan_export_pkg\testB\conanfile.py [2020-04-15 23:43:55,902]
Exporting package recipe
TestB/1.0.0@test/develop exports: Copied 1 '.a' file: liblib4.a
TestB/1.0.0@test/develop: A new conanfile.py version was exported
TestB/1.0.0@test/develop: Folder: C:\Users\dmitry\.conan\data\TestB\1.0.0\test\develop\export
TestB/1.0.0@test/develop: Exported revision: 22339369af2e86a25c2495112798a60f
DEBUG :graph_builder.py[391]: GRAPH: new_node: TestB/1.0.0@test/develop [2020-04-15 23:43:55,916]
DEBUG :graph_builder.py[391]: GRAPH: new_node: TestA/1.0.0@test/develop [2020-04-15 23:44:00,921]
DEBUG :graph_builder.py[66]: GRAPH: Time to load deps 5.00782585144043 [2020-04-15 23:44:00,921]
TestA/1.0.0@test/develop: WARN: Package is corrupted, removing folder: C:\Users\dmitry\.conan\data\TestA\1.0.0\test\develop\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
Traceback (most recent call last):
  File "c:\dev\tools\python3\lib\shutil.py", line 606, in _rmtree_unsafe
    os.unlink(fullname)
PermissionError: [WinError 32] File busy: 'C:\\Users\\dmitry\\.conan\\data\\TestA\\1.0.0\\test\\develop\\package\\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9\\config.txt'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\dev\tools\python3\lib\site-packages\conans\client\command.py", line 1969, in run
    method(args[0][1:])
  File "c:\dev\tools\python3\lib\site-packages\conans\client\command.py", line 939, in export_pkg
    info = self._conan.export_pkg(conanfile_path=args.path,
  File "c:\dev\tools\python3\lib\site-packages\conans\client\conan_api.py", line 81, in wrapper
    return f(api, *args, **kwargs)
  File "c:\dev\tools\python3\lib\site-packages\conans\client\conan_api.py", line 417, in export_pkg
    export_pkg(self.app, recorder, new_ref, source_folder=source_folder,
  File "c:\dev\tools\python3\lib\site-packages\conans\client\cmd\export_pkg.py", line 21, in export_pkg
    deps_graph = graph_manager.load_graph(ref, None, graph_info=graph_info, build_mode=[ref.name],
  File "c:\dev\tools\python3\lib\site-packages\conans\client\graph\graph_manager.py", line 109, in load_graph
    return self._resolve_graph(root_node, graph_info, build_mode, check_updates, update, remotes,
  File "c:\dev\tools\python3\lib\site-packages\conans\client\graph\graph_manager.py", line 230, in _resolve_graph
    deps_graph = self._load_graph(root_node, check_updates, update,
  File "c:\dev\tools\python3\lib\site-packages\conans\client\graph\graph_manager.py", line 340, in _load_graph
    self._recurse_build_requires(graph, builder, check_updates, update, build_mode,
  File "c:\dev\tools\python3\lib\site-packages\conans\client\graph\graph_manager.py", line 277, in _recurse_build_requires
    self._binary_analyzer.evaluate_graph(graph, build_mode, update, remotes, nodes_subset, root)
  File "c:\dev\tools\python3\lib\site-packages\conans\client\graph\graph_binaries.py", line 321, in evaluate_graph
    self._evaluate_node(node, build_mode, update, remotes)
  File "c:\dev\tools\python3\lib\site-packages\conans\client\graph\graph_binaries.py", line 167, in _evaluate_node
    self._process_node(node, pref, build_mode, update, remotes)
  File "c:\dev\tools\python3\lib\site-packages\conans\client\graph\graph_binaries.py", line 209, in _process_node
    metadata = self._evaluate_clean_pkg_folder_dirty(node, package_layout, package_folder, pref)
  File "c:\dev\tools\python3\lib\site-packages\conans\client\graph\graph_binaries.py", line 61, in _evaluate_clean_pkg_folder_dirty
    rmdir(package_folder)  # Do not remove if it is EDITABLE
  File "c:\dev\tools\python3\lib\site-packages\conans\util\files.py", line 248, in rmdir
    shutil.rmtree(path, onerror=_change_permissions)
  File "c:\dev\tools\python3\lib\shutil.py", line 730, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "c:\dev\tools\python3\lib\shutil.py", line 608, in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())
  File "c:\dev\tools\python3\lib\site-packages\conans\util\files.py", line 243, in _change_permissions
    raise OSError("Cannot change permissions for {}! Exception info: {}".format(path, exc_info))
OSError: Cannot change permissions for C:\Users\dmitry\.conan\data\TestA\1.0.0\test\develop\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9\config.txt! Exception info: (<class 'PermissionError'>, PermissionError(13, 'File busy'), <traceback object at 0x00000216A5A91280>)

ERROR: Cannot change permissions for C:\Users\dmitry\.conan\data\TestA\1.0.0\test\develop\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9\config.txt! Exception info: (<class 'PermissionError'>, PermissionError(13, 'File busy'), <traceback object at 0x00000216A5A91280>)
```

Why the folder from the package `testA` is removing during exporting the package `testB`?
```
TestA/1.0.0@test/develop: WARN: Package is corrupted, removing folder: C:\Users\dmitry\.conan\data\TestA\1.0.0\test\develop\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
Traceback (most recent call last):
```

How to completely turn off manipulations with other (required) packages during the command `export-pkg`?

**Comment by memsharded** [comment]
As a general principle, the Conan cache is not concurrent. It is not designed for running parallel processes against it at the same time. It has some provisioning for some operations, but something partial, not a general things.

In your case, if the ``package()`` method of ``TestA`` hasnt finished yet, you cannot export ``TestB`` yet. It requires to resolve the full dependency graph, including ``TestA``, to be able to compute its own ``package_id``. If I understood correctly the traces, it cant, because ``TestA`` file handle is still open.

We need to have a detailed look to the above code and tests, but as a quick feedback, this could be the reason.

**Comment by DmitrySokolov** [comment]
I wonder why the package `testB` tries to delete folders of the package `testA`?

**Comment by DmitrySokolov** [comment]
Why the following sequence is working OK?
1) `conan export testA`
2) N.B. there are still no packaged files of `testA` here
3) `conan export-pkg testB`

**Comment by memsharded** [comment]
Hi @DmitrySokolov 

I have been having a look at your issue and checking your code. I would say that your flow

> 1. conan export testA
> 2. N.B. there are still no packaged files of testA here
> 3. conan export-pkg testB

Is not ok in the general case. To be able to export-pkg testB, you need to have its dependencies already. If testB->testA, in order to build it you need first to have the package testA not only exported but also created (or export-pkg the binary). That binary should be there even before building testB.

So regarding the parallel execution, it is not possible to run the ``conan export-pkg`` of those two testA and testB packages in parallel, as the testA ``export-pkg`` should have been ran even before testB is built. Am I missing something here?

**Comment by DmitrySokolov** [comment]
> The `export-pkg` command let you create a package from **already existing files** in your working folder, it can be useful if you are using **a build process external to Conan** and do not want to provide it with the recipe.

The build artifacts of the package `testB` are already exist, the very goal is just to pack files. So, what is the purpose of checking dependencies of the package `testB` on the "build stage" (N.B. not on the "consume stage")? I expect that the command `export-pkg` should only pack files (comparing to `create`).

Let me explain the situation. I have more than 400 packages (the most of recipes are script-generated). They make up a complex dependency tree. I need somehow to pack them all. So, I run in the loop the command:

```bat
conan export-pkg ...path\package
```

... and get the error: "... can't find dependencies ...". i.e. I need to manually create the dependency three and sort topologically all nodes to get the ordered build list? It is not the thing I want to get from the package manager. I found that the following is working

1. for each package run: `conan export ...path\package`
2. for each package run: `conan export-pkg ...path\package`

The only drawback is the total time. It takes about 2.5 hours (~8000 sec) to pack all packages. I tried to run commands in parallel, and got the total time ~400 sec. But sometimes (as described in the top comment) all crashes.

**Comment by memsharded** [comment]
>The build artifacts of the package testB are already exist, the very goal is just to pack files. So, what is the purpose of checking dependencies of the package testB on the "build stage" (N.B. not on the "consume stage")? I expect that the command export-pkg should only pack files (comparing to create).

The problem is that you need the "package_id" of testB. You know what files you want to pack, but it is necessary to compute where they are packaged (the package_id). That package_id is computed including the dependencies, so you need to compute the full dependency graph of testB, and that includes knowing all the details of testA, down to the package revision of testA, that is computed as the checksum of the binary files in testA final package.

**Comment by DmitrySokolov** [comment]
Sorry, I still did not catch.

> `conan create` executes methods of a conanfile.py in the following order:
> 
> 1. config_options()
> 2. configure()
> 3. requirements()
> 4. package_id()
> 5. build_requirements()
> 6. build_id()
> 7. system_requirements()
> 8. source()
> 9. imports()
> 10. build()
> 11. package()
> 12. package_info()

`package_id` is calculated before `build` and `package` even for `conan export-pkg`, is not it?
And the package payload is not required to calculate package_id, is not it?

I added some logging to the example above:

1. The sequence is `conan export testA ;  conan export-pkg testB`
```bat
DEBUG :export.py      [110]: EXPORT: C:\Dev\Projects\conan-test\conan_export_pkg\testA\conanfile.py [2020-05-05 22:49:24,556]
Exporting package recipe
TestA/1.0.0@test/develop exports: Copied 3 '.a' files: liblib1.a, liblib2.a, liblib3.a
TestA/1.0.0@test/develop: A new conanfile.py version was exported
TestA/1.0.0@test/develop: Folder: C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestA\1.0.0\test\develop\export
TestA/1.0.0@test/develop: Exported revision: fd0f85b0c674fd1ab948fb3b059821cc

DEBUG :export.py      [110]: EXPORT: C:\Dev\Projects\conan-test\conan_export_pkg\testB\conanfile.py [2020-05-05 22:49:25,032]
Exporting package recipe
TestB/1.0.0@test/develop exports: Copied 1 '.a' file: liblib4.a
TestB/1.0.0@test/develop: A new conanfile.py version was exported
TestB/1.0.0@test/develop: Folder: C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestB\1.0.0\test\develop\export
TestB/1.0.0@test/develop: Exported revision: 42541fdb1f7333a8b7cdbf5c0a41c8c0
DEBUG :graph_builder.py[404]: GRAPH: new_node: TestB/1.0.0@test/develop [2020-05-05 22:49:25,046]
TestB/1.0.0@test/develop: --- TestB: 01. config_options ---
TestB/1.0.0@test/develop: --- TestB: 02. configure ---
TestB/1.0.0@test/develop: --- TestB: 03. config_requirements ---
DEBUG :graph_builder.py[404]: GRAPH: new_node: TestA/1.0.0@test/develop [2020-05-05 22:49:30,052]
TestA/1.0.0@test/develop: --- TestA: 01. config_options ---
TestA/1.0.0@test/develop: --- TestA: 02. configure ---
TestA/1.0.0@test/develop: --- TestA: 03. config_requirements ---
DEBUG :graph_builder.py[67]: GRAPH: Time to load deps 5.010640382766724 [2020-05-05 22:49:30,054]
TestA/1.0.0@test/develop: --- TestA: 04. package_id ---
DEBUG :rest_client_common.py[156]: REST: ping: https://conan.bintray.com/v1/ping [2020-05-05 22:49:30,057]
DEBUG :rest_client.py [58]: REST: Cached capabilities for the remote: ['complex_search', 'checksum_deploy', 'revisions'] [2020-05-05 22:49:30,846]
DEBUG :rest_client_common.py[183]: REST: get: https://conan.bintray.com/v1/conans/TestA/1.0.0/test/develop/packages/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/download_urls [2020-05-05 22:49:30,848]
DEBUG :rest_client_common.py[29]: REST ERROR: <class 'conans.errors.NotFoundException'> [2020-05-05 22:49:31,044]
TestB/1.0.0@test/develop: --- TestB: 04. package_id ---
TestB/1.0.0@test/develop: Forced build from source
Packaging to cbba87c4ec83771af3a2c3c6f78e1520a8de6a01
TestB/1.0.0@test/develop: Generating the package
TestB/1.0.0@test/develop: Package folder C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestB\1.0.0\test\develop\package\cbba87c4ec83771af3a2c3c6f78e1520a8de6a01
TestB/1.0.0@test/develop: Calling package()
TestB/1.0.0@test/develop: --- TestB: 11. package ---
TestB/1.0.0@test/develop: WARN: This conanfile has no package step
DEBUG :packager.py    [123]: PACKAGE: Creating config files to C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestB\1.0.0\test\develop\package\cbba87c4ec83771af3a2c3c6f78e1520a8de6a01 [2020-05-05 22:49:31,053]
TestB/1.0.0@test/develop package(): Packaged 1 '.a' file: liblib4.a
TestB/1.0.0@test/develop: Package 'cbba87c4ec83771af3a2c3c6f78e1520a8de6a01' created
TestB/1.0.0@test/develop: Created package revision 17facd9dae96d15108f4b47ee0f8e681
```
TestA: package_id = 5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
TestB: package_id = cbba87c4ec83771af3a2c3c6f78e1520a8de6a01

But there was REST ERROR while calculating TestA package_id. Is it calculated correctly?

2. The sequence is `conan export-pkg testA ;  conan export-pkg testB`
```bat
DEBUG :export.py      [110]: EXPORT: C:\Dev\Projects\conan-test\conan_export_pkg\testA\conanfile.py [2020-05-05 22:49:31,978]
Exporting package recipe
TestA/1.0.0@test/develop exports: Copied 3 '.a' files: liblib1.a, liblib2.a, liblib3.a
TestA/1.0.0@test/develop: A new conanfile.py version was exported
TestA/1.0.0@test/develop: Folder: C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestA\1.0.0\test\develop\export
TestA/1.0.0@test/develop: Exported revision: fd0f85b0c674fd1ab948fb3b059821cc
DEBUG :graph_builder.py[404]: GRAPH: new_node: TestA/1.0.0@test/develop [2020-05-05 22:49:31,993]
TestA/1.0.0@test/develop: --- TestA: 01. config_options ---
TestA/1.0.0@test/develop: --- TestA: 02. configure ---
TestA/1.0.0@test/develop: --- TestA: 03. config_requirements ---
DEBUG :graph_builder.py[67]: GRAPH: Time to load deps 0.005573272705078125 [2020-05-05 22:49:31,996]
TestA/1.0.0@test/develop: --- TestA: 04. package_id ---
TestA/1.0.0@test/develop: Forced build from source
Packaging to 5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
TestA/1.0.0@test/develop: Generating the package
TestA/1.0.0@test/develop: Package folder C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestA\1.0.0\test\develop\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
TestA/1.0.0@test/develop: Calling package()
TestA/1.0.0@test/develop: --- TestA: 11. package ---
TestA/1.0.0@test/develop: WARN: This conanfile has no package step
DEBUG :packager.py    [123]: PACKAGE: Creating config files to C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestA\1.0.0\test\develop\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 [2020-05-05 22:49:42,007]
TestA/1.0.0@test/develop package(): Packaged 1 '.txt' file: config.txt
TestA/1.0.0@test/develop package(): Packaged 3 '.a' files: liblib1.a, liblib2.a, liblib3.a
TestA/1.0.0@test/develop: Package '5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9' created
TestA/1.0.0@test/develop: Created package revision 7fcc969928449dc3ad15301e7987bc48

DEBUG :export.py      [110]: EXPORT: C:\Dev\Projects\conan-test\conan_export_pkg\testB\conanfile.py [2020-05-05 22:49:42,481]
Exporting package recipe
TestB/1.0.0@test/develop exports: Copied 1 '.a' file: liblib4.a
TestB/1.0.0@test/develop: A new conanfile.py version was exported
TestB/1.0.0@test/develop: Folder: C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestB\1.0.0\test\develop\export
TestB/1.0.0@test/develop: Exported revision: 42541fdb1f7333a8b7cdbf5c0a41c8c0
DEBUG :graph_builder.py[404]: GRAPH: new_node: TestB/1.0.0@test/develop [2020-05-05 22:49:42,493]
TestB/1.0.0@test/develop: --- TestB: 01. config_options ---
TestB/1.0.0@test/develop: --- TestB: 02. configure ---
TestB/1.0.0@test/develop: --- TestB: 03. config_requirements ---
DEBUG :graph_builder.py[404]: GRAPH: new_node: TestA/1.0.0@test/develop [2020-05-05 22:49:47,500]
TestA/1.0.0@test/develop: --- TestA: 01. config_options ---
TestA/1.0.0@test/develop: --- TestA: 02. configure ---
TestA/1.0.0@test/develop: --- TestA: 03. config_requirements ---
DEBUG :graph_builder.py[67]: GRAPH: Time to load deps 5.011693954467773 [2020-05-05 22:49:47,502]
TestA/1.0.0@test/develop: --- TestA: 04. package_id ---
TestB/1.0.0@test/develop: --- TestB: 04. package_id ---
TestB/1.0.0@test/develop: Forced build from source
Packaging to cbba87c4ec83771af3a2c3c6f78e1520a8de6a01
TestB/1.0.0@test/develop: Generating the package
TestB/1.0.0@test/develop: Package folder C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestB\1.0.0\test\develop\package\cbba87c4ec83771af3a2c3c6f78e1520a8de6a01
TestB/1.0.0@test/develop: Calling package()
TestB/1.0.0@test/develop: --- TestB: 11. package ---
TestB/1.0.0@test/develop: WARN: This conanfile has no package step
DEBUG :packager.py    [123]: PACKAGE: Creating config files to C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestB\1.0.0\test\develop\package\cbba87c4ec83771af3a2c3c6f78e1520a8de6a01 [2020-05-05 22:49:47,512]
TestB/1.0.0@test/develop package(): Packaged 1 '.a' file: liblib4.a
TestB/1.0.0@test/develop: Package 'cbba87c4ec83771af3a2c3c6f78e1520a8de6a01' created
TestB/1.0.0@test/develop: Created package revision 17facd9dae96d15108f4b47ee0f8e681
```
There were no any errors, so all was calculated correctly.

TestA: package_id = 5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
TestB: package_id = cbba87c4ec83771af3a2c3c6f78e1520a8de6a01


So, why does it require the binary files of TestA, since the package_id is the same when there are binary files, or when there are no binary files?

**Comment by DmitrySokolov** [comment]
When I add `build_policy = "always"` to recipes, all begin to work, except:
```
ERROR: No package matching X pattern
```
But packages are created, and there is no lock issue.

**Comment by DmitrySokolov** [comment]
When I replace `conan export-pkg` with `conan create`, all works OK
```
DEBUG :export.py      [110]: EXPORT: C:\Dev\Projects\conan-test\conan_export_pkg\testB\conanfile.py [2020-05-05 23:46:19,389]
Exporting package recipe
TestB/1.0.0@test/develop exports: Copied 1 '.a' file: liblib4.a
TestB/1.0.0@test/develop: The stored package has not changed
TestB/1.0.0@test/develop: Exported revision: a824b213fd70ec36b1d468c8b8ed275a
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=Visual Studio
compiler.runtime=MD
compiler.version=16
os=Windows
os_build=Windows
[options]
[build_requires]
[env]

DEBUG :graph_builder.py[404]: GRAPH: new_node: TestB/1.0.0@test/develop [2020-05-05 23:46:19,400]
TestB/1.0.0@test/develop: --- TestB: 01. config_options ---
TestB/1.0.0@test/develop: --- TestB: 02. configure ---
TestB/1.0.0@test/develop: --- sleep(5)
TestB/1.0.0@test/develop: --- TestB: 03. config_requirements ---
TestA/1.0.0@test/develop is locked by another concurrent conan process, wait...
If not the case, quit, and do 'conan remove --locks'
DEBUG :graph_builder.py[404]: GRAPH: new_node: TestA/1.0.0@test/develop [2020-05-05 23:46:29,664]
TestA/1.0.0@test/develop: --- TestA: 01. config_options ---
TestA/1.0.0@test/develop: --- TestA: 02. configure ---
TestA/1.0.0@test/develop: --- TestA: 03. config_requirements ---
DEBUG :graph_builder.py[67]: GRAPH: Time to load deps 10.26831603050232 [2020-05-05 23:46:29,665]
TestA/1.0.0@test/develop: --- TestA: 04. package_id ---
TestB/1.0.0@test/develop: --- TestB: 04. package_id ---
TestB/1.0.0@test/develop: Forced build from source
TestB/1.0.0@test/develop: --- TestB: 05. build_requirements ---
Installing package: TestB/1.0.0@test/develop
Requirements
    TestA/1.0.0@test/develop from local cache - Cache
    TestB/1.0.0@test/develop from local cache - Cache
Packages
    TestA/1.0.0@test/develop:3fb49604f9c2f729b85ba3115852006824e72cab - Cache
    TestB/1.0.0@test/develop:f07027c6eed318e6c75daeac08270ccf0433dfbb - Build

Installing (downloading, building) binaries...
TestA/1.0.0@test/develop: Already installed!
TestA/1.0.0@test/develop: --- TestA: 12. package_info ---
TestB/1.0.0@test/develop: --- TestB: 07. system_requirements ---
TestB/1.0.0@test/develop: --- TestB: 06. build_id ---
TestB/1.0.0@test/develop: Configuring sources in C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestB\1.0.0\test\develop\source
TestB/1.0.0@test/develop: --- TestB: 08. source ---
TestB/1.0.0@test/develop: Copying sources to build folder
DEBUG :installer.py   [115]: BUILD: Copied to C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestB\1.0.0\test\develop\build\f07027c6eed318e6c75daeac08270ccf0433dfbb [2020-05-05 23:46:29,685]
DEBUG :installer.py   [116]: BUILD: Files copied liblib4.a [2020-05-05 23:46:29,686]
TestB/1.0.0@test/develop: Building your package in C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestB\1.0.0\test\develop\build\f07027c6eed318e6c75daeac08270ccf0433dfbb
INFO  :installer.py   [120]: GENERATORS: Writing generators [2020-05-05 23:46:29,688]
TestB/1.0.0@test/develop: Generator txt created conanbuildinfo.txt
TestB/1.0.0@test/develop: --- TestB: 09. imports ---
DEBUG :build.py       [12]: Call conanfile.build() with files in build folder: ['conanbuildinfo.txt', 'liblib4.a'] [2020-05-05 23:46:29,689]
TestB/1.0.0@test/develop: Calling build()
TestB/1.0.0@test/develop: --- TestB: 10. build ---
TestB/1.0.0@test/develop: WARN: This conanfile has no build step
TestB/1.0.0@test/develop: Package 'f07027c6eed318e6c75daeac08270ccf0433dfbb' built
TestB/1.0.0@test/develop: Build folder C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestB\1.0.0\test\develop\build\f07027c6eed318e6c75daeac08270ccf0433dfbb
TestB/1.0.0@test/develop: Generated conaninfo.txt
TestB/1.0.0@test/develop: Generated conanbuildinfo.txt
TestB/1.0.0@test/develop: Generating the package
TestB/1.0.0@test/develop: Package folder C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestB\1.0.0\test\develop\package\f07027c6eed318e6c75daeac08270ccf0433dfbb
TestB/1.0.0@test/develop: Calling package()
TestB/1.0.0@test/develop: --- TestB: 11. package ---
TestB/1.0.0@test/develop: WARN: This conanfile has no package step
DEBUG :packager.py    [123]: PACKAGE: Creating config files to C:\Dev\Projects\conan-test\conan_export_pkg\.conan\data\TestB\1.0.0\test\develop\package\f07027c6eed318e6c75daeac08270ccf0433dfbb [2020-05-05 23:46:29,703]
TestB/1.0.0@test/develop package(): Packaged 1 '.a' file: liblib4.a
TestB/1.0.0@test/develop: Package 'f07027c6eed318e6c75daeac08270ccf0433dfbb' created
TestB/1.0.0@test/develop: Created package revision d4bd86dcafa23156be47edb5b57bf868
TestB/1.0.0@test/develop: --- TestB: 11. package_info ---
```
Even the locks is handled correctly:
```
TestA/1.0.0@test/develop is locked by another concurrent conan process, wait...
If not the case, quit, and do 'conan remove --locks'
```

**Comment by memsharded** [comment]
When the ``package_id()`` is evaluated, the dependency graph has been computed, and the upstream packages are known. The ``build()`` and ``package()`` method of the recipe that implements that ``package_id()`` are executed after it, but the ``build()`` and ``package()`` methods of the dependencies must have ran before the consumer ``package_id()``.

The fact that the actual value of package_id might or might not change, is because the different ``package_id_modes``. If you change the ``default_package_id_mode=package_revision_mode``, that will take into account the dependencies package revisions, which is the hash of the contents of the binaries in the package. So to compute the ``package_id()`` the actual binaries of the dependencies need to be known, and need to be packaged.

It might be ok if there is not binary if the ``package_id_mode`` is not using the dependency TestA package revision (hash of the binaries), if the package is not available, as long as the recipe of TestA contains the necessary information to compute the TestB packageID. But this is done in separate steps, first the dependencies of TestB are gathered, then the package_id of TestB is computed. What is not allowed is to be modifying concurrently the TestA package just at the same moment that the TestB ``export-pkg`` is gathering its information to compute its ``package_id``.

I wonder, why don't you start those ``export-pkg`` sooner? As soon as you build the relevant thing. Which always be earlier than the consumers. In this case, as soon as you build locally the TestA binaries, you could be launching that export-pkg, while you start building TestB, and then ``export-pkg`` TestB when you are done. That would do virtually impossible to have these race conditions, wouldn't it? Otherwise, the ``conan graph build-order`` command would be able to give a list of lists that define the order in which operations can be done, including which ones can be done in parallel, because they are mutually independent.

**Comment by DmitrySokolov** [comment]
I have a huge third-party library (like Boost) that consists of hundreds of modules. It has own build system. After the build completes I get the folder with binary files that I need to transform to Conan packages. Another thing I have is the directory tree with Conan recipes. So, I need somehow run `conan export-pkg` for all these recipes.

---

> Otherwise, the conan graph build-order command would be able to give a list of lists that define the order 

Thanks, I will look at it.

**Comment by memsharded** [comment]
At the moment there is nothing we can do in the Conan client. The current cache design is not concurrent, and this is failing due to race conditions because of concurrent export-pkgs. Making the cache concurrent will be a huge effort, at some point  (maybe while or after Conan 2.0) we will consider a cache redesign with some other approach that will tackle concurrency.

**Comment by memsharded** [comment]
We didn't have the time to make the cache concurrent for 2.0, the changes to admit multiple revisions were already huge, so we will have to start concurrency in the new cache from scratch. Moving this to 2.X scope.

**Comment by memsharded** [comment]
Closing in favor of https://github.com/conan-io/conan/issues/15840 where the Conan 2 cache concurrency future progress will be reported.
